Add a validating webhook for ingress sanity check

tjamet opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
* Add a feature to reject the admission of ingresses that generates invalid configuration

Running nginx ingress controller in production for a while, we are pretty happy about it, but we occasionally have an issue where an ingress with a corrupted configuration is injected, and prevents from any other future update of the ingress nginx configuration.
Subsequently, we need to find out the ingress causing problems and fix it.
This PR increases the feedback loop by allowing, optionally, to reject the ingress, before it gets inserted and hence preserves the sanity of the configuration.

Once the patch is applied, and the controller configured, this is the user experience in case of an invalid ingress:

![image](https://user-images.githubusercontent.com/6061440/53275936-5c057c80-36fd-11e9-94ef-8e9d5ea8753e.png)


**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #2833
fixes #3218
fixes #3435
fixes #3459

**Special notes for your reviewer**:
I am aware that this PR needs testing and documentation.
In case the option of adding this feature is accepted, I will as well contribute documentation and tests.

aledbf commented:
@tjamet first, thank you for working on this!
That said:
- please rebase
- to update the dependencies please run:
  - rm -rf images/grpc-fortune-teller/ (you can restore that before the commit)
  - make dep-ensure

aledbf commented:
@tjamet as you said, we need tests for this :)
What's the behavior when you have multiple deployments with different classes?

aledbf commented on internal/ingress/controller/controller.go:
```diff
@@ -191,6 +197,35 @@ func (n *NGINXController) syncIngress(interface{}) error {
+	ings := n.store.ListIngresses()
+	ings = append(ings, &ingress.Ingress{*ing, annotations.NewAnnotationExtractor(n.store).Extract(ing)})
+
+	// sort Ingresses using the CreationTimestamp field
```
This is not required because ListIngresses alredy sorts the ingresses

tjamet commented:
@aledbf thanks for the fast feedback
I am currently rebasing

Indeed I will then work on tests and documentation!
Regarding multiple deployments, the expected behavior is to accept ingresses that do not [match the class](https://github.com/kubernetes/ingress-nginx/pull/3802/commits/c07f55330b6bb00128ede5a80d4d5c279b6015a9#diff-86ac9ff9d75a0c5005c116e766a6127dR203). This said, I need further tests with multiple deployments

aledbf commented:
> Regarding multiple deployments, the expected behavior is to accept ingresses that do not [match the class](https://github.com/kubernetes/ingress-nginx/pull/3802/commits/c07f55330b6bb00128ede5a80d4d5c279b6015a9#diff-
86ac9ff9d75a0c5005c116e766a6127dR203).

Yes, I saw that but I am not sure about the webhook behavior with multiple deployments

ElvinEfendi commented:
Also fixes https://github.com/kubernetes/ingress-nginx/issues/3588

ElvinEfendi commented:
@tjamet thanks for working on this! It'd be great if you describe your approach, discuss how much overhead this is adding to the core event loop, and why would I not want to have this always on.

tjamet commented:
@aledbf I am seeing so far 2 parameters that can change the set of ingresses considered for configuration generation in a given controller:
- `--ingress-class`
- `--watch-namespace`
I am hence suspecting multiple deployments to have 4 different ways to consider or not an ingress.

Following your comment, I added the filter based on the namespace name as well for to behave consistently with the deployment. This should handle all combinations of `--ingress-class` and `--watch-namespace`, as soon as the validating webhook is enabled for all deployments.

The approach @ElvinEfendi  is pretty simple in the end, it consists of exposing, outside of the core event loop an HTTP handler, compliant with Kubernetes validating webhook interface, to ensure any ingress that is about to be inserted in the cluster would generate a syntactically valid configuration **_together with the rest of the known ingresses_** using the pre-existing features. This handler should be exposed through a service to a ValidatingWebhookConfiguration object.

As this handler is run outside of the core event loop, I have tested a benchmark that shows the following results (reported by `go test -bench '.' -benchmem -run Benchmark -v` within the `./bin/go-in-docker.sh`):

|number of pre existing ingresses|(go) memory used per request|time per validating request|
|-|-|-|
|0|136342 B/op|25926680 ns/op|
|20|637230 B/op|26366018 ns/op|
|50|1415960 B/op|38304690 ns/op|
|100|2587819 B/op|47551526 ns/op|
|200|5566906 B/op|71309565 ns/op|
|500|12790632 B/op|128302310 ns/op|
|1000|28805841 B/op|226466720 ns/op|

Because the request is served through a service to using the ingress controller as a backend, webhook requests are load-balanced through all ingress controllers and hence reduces the load on a single pod in case multiple ingresses are inserted at the same time.

Another option to get better performances could be to drop the test with all pre-existing ingresses, which could probably be good enough to detect most configuration errors (syntax errors mainly).

Regarding the default enabling of the feature, I would tend to, as a system operator, to recommend to enable it systematically.
Although, it looks to me difficult to enable it by default directly after implementation. This would change the permissions the ingress controller would require to grant it system wide privileges to inject the `ValidatingWebhookConfiguration` object.
Also the way ValidatingWebhook works, they require a self signed certificate generated for the correct `<service>.<namespace>.svc` domain, which would be more difficult to integrate out of the box inside the ingress controller. My approach would rather be to make it easy in the chart to use such a feature to start with. Happy to read other opinions or concerns

tjamet commented:
> We can provide a script like https://gist.github.com/aledbf/11b3b4f882c632d021532edda8a5dc52 (I need to cleanup that) to use the feature already available in k8s to generate such certificate

Sure!
I can add it to this PR
We have it deployed using helm, using:
```
{{- $cn := printf "%s.%s.svc" ( include "nginx-ingress.validatingWebhook.fullname" . ) .Release.Namespace }}
{{- $ca := genCA (printf "%s-ca" ( include "nginx-ingress.validatingWebhook.fullname" . )) .Values.validatingWebhook.certificateValidity -}}
{{- $cert := genSignedCert $cn nil nil .Values.validatingWebhook.certificateValidity $ca -}}
```
This could also be integrated afterwards to the helm chart, that I could contribute to as well

Regarding impact on resources, here are graphs, with the deployment in the middle.
The previous version was `0.19.0`
![screenshot 2019-03-06 at 16 40 34](https://user-images.githubusercontent.com/6061440/54078074-46c64b80-42c2-11e9-8e9e-2550e16dc051.png)

aledbf commented:
> This could also be integrated afterwards to the helm chart, that I could contribute to as well

Just in case, we also need the plain yaml files, like all the files in https://github.com/kubernetes/ingress-nginx/tree/master/deploy

aledbf commented:
@tjamet is anything I can do to help with this PR?

tjamet commented:
@aledbf
I need to add some comments to describe the new methods to handle
certificate rotation, I am planning to do it as soon as I have a stable
internet connection, being on a trip for one week more
I will also update documentation about the certificate generation
Regarding the plain yamls, as it needs the secrets, I added sample
templates, if you have a better idea, I am taker
Happy to update it with anything else required, let me know
If you need to update it meanwhile I have a weak internet connection,
please do


On Thu, 14 Mar 2019 at 18:51, Manuel Alejandro de Brito Fontes <
notifications@github.com> wrote:

> @tjamet <https://github.com/tjamet> is anything I can do to help with
> this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/ingress-nginx/pull/3802#issuecomment-472844140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFx9gILBzTGA7rhexMrcSxcy4FUSFaiOks5vWkj1gaJpZM4bL6P->
> .
>

zq-david-wang commented:
Just a silly question, would this validating webhook reject ingress resource that are valid for ingress controller of other kind? For example, the regular expression syntax for  traefik is weird and would cause nginx ingress controller crash, if I deploy nginx ingress controller with this validating webhook, can I use traefik within the same cluster? (To be honest, the metrics endpoint for traefik is way more useful than nginx's)

tjamet commented:
Hi @zq-david-wang 
> would this validating webhook reject ingress resource that are valid for ingress controller of other kind

The validating webhook is filtering on the `kubernetes.io/ingress.class`, ingress resources that match a different class will be accepted.
In case you run multiple ingress controllers (we do run several nginx ingress controllers) in a cluster, I would expect to find controllers deployed with different classes, hence ingresses for different ingresses kind (traefik in the case you are talking about) will be accepted.

In case both ingresses are run with the same class, both ingress controllers would generate a configuration for the ingress and hence the validating webhook will check the validity of the generated configuration

tjamet commented:
Hi,
I think I addressed the concerns, are there any other one I need to address? Or is it good enough for a first version?
I will be happy to keep having a look at it and improve it with other ideas later on

aledbf commented:
@tjamet please rebase and squash the commits

tjamet commented:
Rebased and commits squashed

aledbf commented:
@tjamet please rebase. After that, I plan to build a dev image to test and merge this PR

ElvinEfendi commented:
/lgtm

ElvinEfendi commented:
Thanks @tjamet!

towolf commented:
This is really cool. Thanks @tjamet.

dcvtruong commented:
Hi,

How do I retrieve and apply the fix for the issue in version 0.21? If someone could provide the location.

Thank you for the fix.
